urldeb33fd2fa_friedemann_distripedia_tgz-pJ8-org_json_TestJ8/src/grinder/controller/CtrlServerThread.java

```diff
@@ -0,0 +1,359 @@
+	 * @param socket bound client socket
+	 */
+	public CtrlServerThread(final Controller ctrl, final CtrlServer server,
+			final String name, final Socket socket) {
+		super(name);
+
+		this.ctrl = ctrl;
+		this.server = server;
+		clientSocket = socket;
+		this.name = name;
+
+		ctrl.cgui.logMessage("<" + name + "> server connection thread initialized");
+	}
+
+	/**
+	 * Write to object output stream.
+	 *
+	 * @param obj the obj
+	 */
+	public void writeToOutputStream(final Object objOut) {
+		if (oos != null)
+		{
+			try
+			{
+				oos.writeObject(objOut);
+				oos.flush();
+
+			} catch (final IOException e)
+			{
+				ctrl.cgui.logMessage("<" + name + "> Can't write object to output stream! "
+						+ e);
+				throw new RuntimeException("output stream fail");
+			}
+		} else
+		{
+			ctrl.cgui.logMessage("<" + name
+					+ "> Can't write object to output stream! (oos==null)");
+		}
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see java.lang.Thread#run()
+	 */
+	@Override
+	public void run() {
+
+		try
+		{
+			// set up in and output streams
+			ois = new ObjectInputStream(clientSocket.getInputStream());
+			oos = new ObjectOutputStream(clientSocket.getOutputStream());
+			ctrl.cgui.logMessage("<" + name + "> server iostreams set up");
+
+			requestClientInformation();
+
+			// listen for incoming data
+			while (serverThreadListening)
+			{
+
+				/* *********************************************
+				 * the client generally only sends messages about his state - only
+				 * exceptions: 1) when the server asks for the log file 2) when the
+				 * client receives a test ticket
+				 * *********************************************
+				 */
+
+				ClientState cstate = null;
+
+				try
+				{
+					cstate = (ClientState) ois.readObject();
+					ctrl.cgui.logMessage("<" + name + "> in: <" + cstate.toString() + ">");
+					clientState = cstate;
+					ctrl.cgui.updateClientTableData();
+
+				} catch (final ClassCastException e)
+				{
+					ctrl.cgui
+							.logMessage("(�`�._.�(�`�._.receiving alien signals._.���)�._.���)");
+				}
+
+				switch (cstate)
+				{
+				case RECEIVE_TICKET_OK:
+					deployedTicketID = (String) ois.readObject();
+					ctrl.cgui.updateClientTableData();
+					break;
+
+				case ARMED:
+					break;
+
+				case CLEANUP:
+					deployedTicketID = "";
+					ctrl.cgui.updateClientTableData();
+					break;
+
+				case TESTING:
+					break;
+
+				case FINISHED:
+					ctrl.checkForTestFinish();
+					break;
+
+				case SENDING_LOGFILE:
+
+					initializeProgressMonitor();
+
+					long bytesReceived = 0;
+					final long bytesTotal = ois.readLong(); // get size of total
+					// bytes
+					// from client
+
+					// open file to write to
+					final String fileName = GrinderSettings.CTRL_LOG_FILE_FOLDER
+							+ deployedTicketID + ".log";
+					final File completeLogFile = new File(fileName);
+
+					// append data if file exists
+					final FileOutputStream fos = new FileOutputStream(completeLogFile, true);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fos*. The resource is referenced at lines 190, 217. The resource is closed at line 218. There are other execution paths that do not contain closure statements, for example, when InputStream.read throws an exception. Ensure *fos* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).